packages/clients/src/aptos/wallet.ts

```diff
@@ -0,0 +1,431 @@
+   * @param run The code to run.
+   * @returns Result of `run`.
+   */
+  async withToken<T>(run: (token: string) => Promise<T>): Promise<T> {
+    if (!this.token) {
+      return await run(await this.refreshToken());
+    } else {
+      try {
+        return await run(this.token);
+      } catch (e: unknown) {
+        if (e instanceof JSONRPCError && e.code === -32602) {
+          const details = errorDetails(e);
+          if (details?.details?.includes("Bad session token")) {
+            return await run(await this.refreshToken());
+          }
+        }
+        throw e;
+      }
+    }
+  }
+}
+
+/**
+ * A signer based on Shinami's invisible wallet for Aptos.
+ *
+ * It transparently manages session token refreshes.
+ */
+export class ShinamiWalletSigner {
+  readonly walletId: string;
+  readonly walletClient: WalletClient;
+  private readonly session: KeySession;
+
+  private address?: string;
+
+  constructor(
+    walletId: string,
+    walletClient: WalletClient,
+    session: KeySession,
+  );
+  constructor(
+    walletId: string,
+    walletClient: WalletClient,
+    secret: string,
+    keyClient: KeyClient,
+  );
+  constructor(
+    walletId: string,
+    walletClient: WalletClient,
+    secretOrSession: string | KeySession,
+    keyClient?: KeyClient,
+  ) {
+    this.walletId = walletId;
+    this.walletClient = walletClient;
+
+    if (secretOrSession instanceof KeySession) {
+      this.session = secretOrSession;
+    } else {
+      if (!keyClient) throw new Error("Must provide keyClient with secret");
+      this.session = new KeySession(secretOrSession, keyClient);
+    }
+  }
+
+  /**
+   * Retrieves the wallet address.
+   * @param autoCreate Whether to automatically create the wallet (off chain) if it doesn't exist yet.
+   *    If `false`, and the wallet doesn't exist, an error will be thrown.
+   * @param onChain whether to initialize the address on chain after creation. It will use the network
+   *    attached to the access key in the sessionToken. Only relevant if autoCreate is set to `true`
+   * @returns Wallet address.
+   */
+  async getAddress(autoCreate = false, onChain = false): Promise<string> {
+    if (!this.address)
+      this.address = await this._getAddress(autoCreate, onChain);
+    return this.address;
+  }
+
+  private async _getAddress(
+    autoCreate: boolean,
+    onChain: boolean,
+  ): Promise<string> {
+    try {
+      return (await this.walletClient.getWallet(this.walletId)).accountAddress;
+    } catch (e: unknown) {
+      if (e instanceof JSONRPCError && e.code === -32602 && autoCreate) {
+        const address = await this.tryCreate(onChain);
+        if (address) return address;
+        return (await this.walletClient.getWallet(this.walletId))
+          .accountAddress;
+      }
+      throw e;
+    }
+  }
+
+  /**
+   * Tries to create this wallet if it doesn't exist.
+   * @param onChain If set, it will try to create the wallet and initialize it on chain as well.
```

Comment:
Clarify this must be called with an access key with gas auth, and we'll draw from the fund.